serverless/serverless
# Fix ImportValue handling in existing S3 buckets #6416

## What did you implement:

Closes #6416



## How did you implement it:

Currently, bucket ARN for the permissions is being generated using string concatenation, even if the bucket name is not a string.

I changed the policy generating code to use `Fn::Join` if the bucket name is not a string but a CloudFormation function (or anything other than a string).

If the bucket name is a string, then the ARN is still being generated using string concatenation on Serverless side (for backwards compatibility and to keep the template simpler)



## How can we verify it:



* Create an S3 bucket in a CloudFormation stack, exporting its name
* Use this template:

```
service: test

provider:
  name: aws
  runtime: nodejs10.x

functions:
  hello:
    handler: handler.hello
    events:
      - s3:
          bucket: !ImportValue test
          events:
            - s3:ObjectCreated:*
          existing: true
```
, substituting the exported value from the previously created stack
* Make sure the new stack compiles and deploys.

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

       **Validate via `npm test`**
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_

**_Is this ready for review?:_** YES  
**_Is it a breaking change?:_** NO
